solidity/random-beacon/contracts/RandomBeacon.sol

```diff
@@ -548,6 +589,39 @@ contract RandomBeacon is Ownable {
         }
     }
 
+    /// @notice Punishes the given operators by kicking them out from the
+    ///         sortition pool and blocking their re-join for a given period
+    ///         of time.
+    /// @dev By the way, this function releases gas deposits made by operators
+    ///      during their registration. See `registerOperator` function. This
+    ///      action makes punishments cheaper gas-wise.
+    /// @param operators Addresses of punished operators.
+    /// @param punishmentDuration Duration of the punishment period in seconds.
+    function punishOperators(
+        address[] memory operators,
+        uint256 punishmentDuration
+    ) internal {
+        for (uint256 i = 0; i < operators.length; i++) {
+            address operator = operators[i];
+
+            // Set the punishment regardless the operator is actually registered
+            // in the sortition pool.
+            /* solhint-disable not-rely-on-time */
+            // slither-disable-next-line reentrancy-benign
+            punishedOperators[operator] = block.timestamp + punishmentDuration;
+            /* solhint-enable not-rely-on-time */
+
+            // Perform operations on sortition pool only in case the operator
+            // is registered. Otherwise, the punishment set above will prevent
+            // the operator to join the pool in future.
+            if (sortitionPool.isOperatorInPool(operator)) {
+                gasStation.releaseGas(operator);
```

Comment:
How do we know that the total gas refund of `gasStation.releaseGas(operator)` covers used gas in `sortitionPool.removeOperator(operator)`?

Reply:
Well, we don't know at this point. I'm also not sure we should care if it covers the whole gas used in `sortitionPool.removeOperator(operator)`. The goal of the gas station pattern used here is to make `submitRelayEntry` call as cheap as possible for the submitter, especially in case that transaction needs to remove multiple group members from the sortition pool by the way. Currently, we reserve one slot of memory during operator registration. Once we have a real sortition pool in tests (or even better - system tests) we will be able to assess gas consumption and try to optimize by, for instance, reserving more slots upon operator registration. At the same time, we will need to make sure operator registration doesn't become too expensive due to the gas deposit.

Reply:
Ok, thx for the explanation. Can we add a comment somewhere that this will need to be adjusted? Maybe inside the `GasStation` contract?

Reply:
See https://github.com/keep-network/keep-core/pull/2671/commits/210861513bf978a98bd737e3b90a27ca838f0d02